framework/master-key-provider-interface.md

```diff
@@ -0,0 +1,106 @@
+[//]: # (Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.)
+[//]: # (SPDX-License-Identifier: CC-BY-SA-4.0)
+
+# Master Key Provider Interface
+
+## Version
+
+0.1.0-preview
+
+## Implementations
+
+- [Java](https://github.com/aws/aws-encryption-sdk-java/blob/master/src/main/java/com/amazonaws/encryptionsdk/MasterKeyProvider.java)
+- [Python](https://github.com/aws/aws-encryption-sdk-python/blob/master/src/aws_encryption_sdk/key_providers/base.py)
+
+## Overview
+
+Master key providers are responsible for managing [master keys](./master-key-interface.md),
+determining which master keys should be used for encryption,
+and for decrypting data keys using available master keys.
+
+### Legacy
+
+This is a legacy specification.
+Master key providers SHOULD NOT be included in any additional implementations.
+Any new implementations MUST include [Keyrings](./keyring-interface.md) instead.
+
+### Consistency
+
+This specification defines the common behavior between the two implementations
+that determine the REQUIRED functionality.
+
+Other specifics of behavior and API vary between the two implementations.
+
+## Definitions
+
+### Conventions used in this document
+
+The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted
+as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+### Terms
+
+* Provider ID : A value that identifies a master key provider.
+    This concept is equivalent to "key namespace" for Keyrings.
+* Key ID : A value that identifies a master key
+    within the context of a master key provider.
+    This concept is equivalent to "key name" for Keyrings.
+* Provider Info : The value that is written to a serialized encrypted data key
+    that identifies a master key within the context of a master key provider.
+    This MUST always be equal to the master key's key ID
+    with the exception of the raw AES master key.
+    For a detailed description of this exception,
+    see the [Raw AES Keyring specification](./raw-aes-keyring.md).
+
+## Interface
+
+### Get Master Key
+
+This interface provides a way for a master key provider to return master keys.
+
+An implementation MUST support master key selection by key ID.
+
+An implementation MAY support master key selection by provider info or provider ID.
+
+The output of this interface MUST be a master key.
+
+If the master key provider cannot locate an appropriate master key,
+the call MUST result in an error.
+
+### Get Master Keys For Encryption
+
+This interface provides a way for a master key provider to indicate which master keys
+SHOULD be used for encryption.
```

Comment:
This is an artifact of the evolution of these two clients. Originally this the logic for what to do with the returned MKs was in the client itself, so it would have been a "will". Then, when we made CMMs, we moved the logic into the `DefaultCryptoMaterialsManager`. IMO, this makes it a "SHOULD", because the intention is that a MKP will return all of the MKs that it wants you to use, and the only way that we support using MKs is via the `DefaultCryptoMaterialsManager` (pending `MasterKeyProviderKeyring` spec proposal coming soon).